samples/solutions/stingy_connection.py

```diff
@@ -0,0 +1,275 @@
+                        )
+                        await self.device_client.connect()
+                        self.log_info_and_print("Successfully connected the device client...")
+                        await self.do_all_tasks_and_disconnect()
+                        break
+                    except Exception as e:
+                        self.log_error_and_print("Caught exception while trying to connect...")
+                        if type(e) is exceptions.CredentialError:
+                            self.log_error_and_print(
+                                "Failed to connect the device client due to incorrect or badly formatted credentials..."
+                            )
+                            raise
+                        if self.is_retryable(e):
+                            self.log_error_and_print(
+                                "Failed to connect the device client due to retryable error."
+                                "Sleeping and retrying after some time..."
+                            )
+                            await asyncio.sleep(CONNECTION_ATTEMPT_INTERVAL)
+                        else:
+                            self.log_error_and_print(
+                                "Failed to connect the device client due to not-retryable error...."
+                            )
+                            raise
+            # Sleep for 30 minutes and again collect all messages to send
+            self.log_info_and_print(
+                "Will sleep for {} seconds and then retrieve messages to send....".format(
+                    TIME_BETWEEN_CONNECTIONS
+                )
+            )
+            await asyncio.sleep(TIME_BETWEEN_CONNECTIONS)
+
+    async def memory_stats(self):
+        while True:
+            gc_counts = gc.get_stats()
+
+            self.log_info_and_print("GC stats are:-")
+            for item in gc_counts:
+                self.log_info_and_print("collections -> {}".format(item["collections"]))
+                self.log_info_and_print("collected -> {}".format(item["collected"]))
+                self.log_info_and_print("uncollectable -> {}".format(item["uncollectable"]))
+
+            self.log_info_and_print(
+                "Will sleep for {} seconds and then collect memory stats....".format(
+                    STATISTICS_INTERVAL
+                )
+            )
+            await asyncio.sleep(STATISTICS_INTERVAL)
+
+    @profile
+    async def main(self):
+        conn_str = os.getenv("IOTHUB_DEVICE_CONNECTION_STRING")
+        await self.initiate()
+        await self.create_client(conn_str)
+
+        tasks = [
+            asyncio.create_task(self.connect_with_retry_send_all_and_disconnect()),
+            asyncio.create_task(self.enqueue_message()),
+            asyncio.create_task(self.memory_stats()),
+        ]
+
+        try:
+            done, pending = await asyncio.wait(tasks, return_when=asyncio.FIRST_EXCEPTION)
+            await asyncio.gather(*done)
```

Comment:
>await asyncio.gather(*done) [](http://example.com/codeflow?start=12&length=27)

What happens to pending tasks? From what I can tell, they keep running forever.

Reply:
I included wait for the pending tasks.